### DISCUSSION TITLE
Autopilot functions different in Lazer

### DISCUSSION THREAD

**Discussion by Inconsist** [ppy/osu#25354] [Q&A]
It seems like in lazer the cursor just teleports to the next hintobject. (at least sometimes) And this makes reading very awkward and hard. Can the stable version be implemented into lazer?

https://github.com/ppy/osu/assets/87947227/ba42941d-52f8-4e97-9415-3d42f17c1632

**Comment by peppy** [comment]
I do believe we'll improve autopilot to be more usable. Right now it's doing it's best.

**Comment by shawnthwei** [comment]
For others also searching for this issue: I proposed #28166 for fixing this. I play a lot of Autopilot, and with the change I think the gameplay feels more like what one would expect. It's currently not clear whether the pull request will get accepted or not, but even if not, I thought it'd be worth referencing it here for others that look into it.

**Comment by murryguy** [comment]
> It seems like in lazer the cursor just teleports to the next hintobject. (at least sometimes) And this makes reading very awkward and hard. Can the stable version be implemented into lazer?
> 
>  autopilit.comp.mp4

Do you mean: "hitobject"

**Comment by TaterToes** [comment]
I was wondering if we should do a **different approach instead of replay frames**. Since the goal is just automatic cursor movement, and pausing on a HitObject on time.

**Unsure if I can create a pull request while one is active related to autopilot**, but I made a implementation of Autopilot thats based on distance between cursor and hitobject, and currentTime to start of the HitWindow. This mostly fixes the cursor movement and fixes the tapping issue with Autopilot. Here's an example below with the new implementation, which completely removed replay frames.

https://github.com/user-attachments/assets/fe88f5c2-c041-4f9f-a59d-a96c7afd9924



Of course, just showing the video doesn't mean it's free of bugs and wrong implementations, but I just wanted to show a new approach, and **see if it would be accepted before I would even start a pull request**.

**Reply by peppy** [reply]
You're more than welcome to make a pull request. This is how it is done on stable and how we'd eventually see it being done here. We've used autoplay / replays until now because it was the simplest way to get things working many years ago, but as you can tell it has not aged well with people's requirements.

Please make sure to add test coverage to show that it's capable of correct behaviour, if possible.
